src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,640 @@
+use std::{
+    cell::UnsafeCell,
+    fmt::Debug,
+    pin::Pin,
+    sync::{atomic::AtomicU64, Arc},
+    task::{Context, Poll},
+};
+
+use bytes::BytesMut;
+use futures::{pin_mut, Future};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpStream,
+    time::{Instant, Interval},
+};
+use tracing::debug;
+
+use crate::{read_exact_timeout, write_all_timeout};
+
+use super::{
+    common::ClientTimeoutOptions,
+    error::RpcError,
+    message::{ReqType, RespType},
+    packet::{Decode, Encode, Packet, PacketsKeeper, ReqHeader, RespHeader, REQ_HEADER_SIZE},
+    utils::u64_to_usize,
+};
+
+/// TODO: combine `RpcClientConnectionInner` and `RpcClientConnection`
+struct RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// The TCP stream for the connection.
+    stream: UnsafeCell<TcpStream>,
+    /// Options for the timeout of the connection
+    timeout_options: ClientTimeoutOptions,
+    /// Stream auto increment sequence number, used to mark the request and response
+    seq: AtomicU64,
+    /// The instant of the connection, used to calculate the keep alive timeout
+    clock_instant: Instant,
+    /// Received keep alive timestamp, will mark the valid keep alive response.
+    /// If the keep alive response is not received in right mestamp, the connection will be closed,
+    /// If we receive other response, we will update the received_keepalive_timestamp too, just treat it as a keep alive response.
+    received_keepalive_timestamp: AtomicU64,
+    /// Send packet task
+    packets_keeper: UnsafeCell<PacketsKeeper<P>>,
```

Comment:
packets_keeper no longer need UnsafeCell

Reply:
Maybe it is needed to be modified in `ReceiveHeaderFuture`.